### PR TITLE
refactor visualisation_matrix

### PR DESCRIPTION
**Pull request by DominiqueMakowski**
## Features

@strengejacke @bwiernik @mattansb I'm going to slowly start listing the features and possibilities of the visualisation_matrix function as I refactor it. Let me know what you think good ideas / options / caveats are!

#### Factors / Characters

There's not much you can do with factors. It returns the unique levels (and preserves the class depending on whether it's a character or a factor). Also, arbitrary code and values can be flexibly (and quite forgivingly) passed via the `target` argument.

``` r
modelbased::visualisation_matrix(iris$Species)
#> [1] setosa     versicolor virginica 
#> Levels: setosa versicolor virginica

modelbased::visualisation_matrix(iris$Species, target = "['versicolor', 'virginica']")
#> [1] versicolor virginica 
#> Levels: versicolor virginica

modelbased::visualisation_matrix(iris$Species, target = "x = ['versicolor', 'virginica']")
#> [1] versicolor virginica 
#> Levels: versicolor virginica

modelbased::visualisation_matrix(iris$Species, target = "execute_order66 = c('versicolor', 'virginica')")
#> [1] "versicolor" "virginica"

modelbased::visualisation_matrix(iris$Species, target = "as.factor(c('versicolor'))")
#> [1] versicolor
#> Levels: versicolor
```

<sup>Created on 2021-05-12 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

#### Numerics

For numeric vectors, it returns a spread between the min-max. Aside from arbitrary code / range specifications with `target`, options are available to adjust the `length` (default to 10), but also to not use the raw min/max but rather those of a specified interval method. This can be useful when there are outliers and extremes in the data and people want to visualize the links in the more "dense" area of the data.

``` r
modelbased::visualisation_matrix(mtcars$mpg)
#>  [1] 10.40000 13.01111 15.62222 18.23333 20.84444 23.45556 26.06667 28.67778
#>  [9] 31.28889 33.90000

modelbased::visualisation_matrix(mtcars$mpg, length = 4)
#> [1] 10.40000 18.23333 26.06667 33.90000

modelbased::visualisation_matrix(mtcars$mpg, target = "c(1, 2, 3)")  # Arbitrary code
#> [1] 1 2 3

modelbased::visualisation_matrix(mtcars$mpg, target = "[1, 5]")  # Arbitrary range
#>  [1] 1.000000 1.444444 1.888889 2.333333 2.777778 3.222222 3.666667 4.111111
#>  [9] 4.555556 5.000000

modelbased::visualisation_matrix(mtcars$mpg, range = "hdi", ci = 0.90)  # Default ci = 0.95
#>  [1] 10.40000 12.62222 14.84444 17.06667 19.28889 21.51111 23.73333 25.95556
#>  [9] 28.17778 30.40000

modelbased::visualisation_matrix(mtcars$mpg, range = "IQR", ci = 0.75)  
#>  [1] 14.65000 16.09861 17.54722 18.99583 20.44444 21.89306 23.34167 24.79028
#>  [9] 26.23889 27.68750

modelbased::visualisation_matrix(mtcars$mpg, range = "eti", ci = 0.80)  
#>  [1] 14.34 16.09 17.84 19.59 21.34 23.09 24.84 26.59 28.34 30.09
```

<sup>Created on 2021-05-12 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

#### Single Target 

Naturally, the critical thing is how it runs on dataframes. All the options above work and apply for the "target" column (the one(s) that we want to modulate).

``` r
modelbased::visualisation_matrix(iris, target = "Petal.Length", length = 2, range = "hdi")
#> Visualisation Matrix
#> 
#> Petal.Length | Sepal.Length | Sepal.Width | Petal.Width | Species
#> -----------------------------------------------------------------
#>         1.00 |         5.84 |        3.06 |        1.20 |  setosa
#>         6.10 |         5.84 |        3.06 |        1.20 |  setosa
#> 
#> Maintained constant: Sepal.Length, Sepal.Width, Petal.Width, Species

modelbased::visualisation_matrix(iris, target = "Petal.Length = c(1, 2, 3)")
#> Visualisation Matrix
#> 
#> Petal.Length | Sepal.Length | Sepal.Width | Petal.Width | Species
#> -----------------------------------------------------------------
#>            1 |         5.84 |        3.06 |        1.20 |  setosa
#>            2 |         5.84 |        3.06 |        1.20 |  setosa
#>            3 |         5.84 |        3.06 |        1.20 |  setosa
#> 
#> Maintained constant: Sepal.Length, Sepal.Width, Petal.Width, Species

modelbased::visualisation_matrix(iris, target = "Species")
#> Visualisation Matrix
#> 
#> Species    | Sepal.Length | Sepal.Width | Petal.Length | Petal.Width
#> --------------------------------------------------------------------
#> setosa     |         5.84 |        3.06 |         3.76 |        1.20
#> versicolor |         5.84 |        3.06 |         3.76 |        1.20
#> virginica  |         5.84 |        3.06 |         3.76 |        1.20
#> 
#> Maintained constant: Sepal.Length, Sepal.Width, Petal.Length, Petal.Width

modelbased::visualisation_matrix(iris, target = "Species = ['setosa', 'versicolor']")
#> Visualisation Matrix
#> 
#> Species    | Sepal.Length | Sepal.Width | Petal.Length | Petal.Width
#> --------------------------------------------------------------------
#> setosa     |         5.84 |        3.06 |         3.76 |        1.20
#> versicolor |         5.84 |        3.06 |         3.76 |        1.20
#> 
#> Maintained constant: Sepal.Length, Sepal.Width, Petal.Length, Petal.Width
```

<sup>Created on 2021-05-12 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

The way the "remaining" variables are fixed is also adjustable, and can be set as arbitrary values, code, etc.

``` r
modelbased::visualisation_matrix(iris, target = "Petal.Length", factors = "mode") # Instead of "reference" (in this case it's same)
#> Visualisation Matrix
#> 
#> Petal.Length | Sepal.Length | Sepal.Width | Petal.Width | Species
#> -----------------------------------------------------------------
#>         1.00 |         5.84 |        3.06 |        1.20 |  setosa
#>         1.66 |         5.84 |        3.06 |        1.20 |  setosa
#>         2.31 |         5.84 |        3.06 |        1.20 |  setosa
#>         2.97 |         5.84 |        3.06 |        1.20 |  setosa
#>         3.62 |         5.84 |        3.06 |        1.20 |  setosa
#>         4.28 |         5.84 |        3.06 |        1.20 |  setosa
#>         4.93 |         5.84 |        3.06 |        1.20 |  setosa
#>         5.59 |         5.84 |        3.06 |        1.20 |  setosa
#>         6.24 |         5.84 |        3.06 |        1.20 |  setosa
#>         6.90 |         5.84 |        3.06 |        1.20 |  setosa
#> 
#> Maintained constant: Sepal.Length, Sepal.Width, Petal.Width, Species

modelbased::visualisation_matrix(iris, target = "Petal.Length", numerics = "median") # Instead of mean
#> Visualisation Matrix
#> 
#> Petal.Length | Sepal.Length | Sepal.Width | Petal.Width | Species
#> -----------------------------------------------------------------
#>         1.00 |         5.80 |           3 |        1.30 |  setosa
#>         1.66 |         5.80 |           3 |        1.30 |  setosa
#>         2.31 |         5.80 |           3 |        1.30 |  setosa
#>         2.97 |         5.80 |           3 |        1.30 |  setosa
#>         3.62 |         5.80 |           3 |        1.30 |  setosa
#>         4.28 |         5.80 |           3 |        1.30 |  setosa
#>         4.93 |         5.80 |           3 |        1.30 |  setosa
#>         5.59 |         5.80 |           3 |        1.30 |  setosa
#>         6.24 |         5.80 |           3 |        1.30 |  setosa
#>         6.90 |         5.80 |           3 |        1.30 |  setosa
#> 
#> Maintained constant: Sepal.Length, Sepal.Width, Petal.Width, Species

modelbased::visualisation_matrix(iris, target = "Petal.Length", numerics = 0, factors = "reference")
#> Visualisation Matrix
#> 
#> Petal.Length | Sepal.Length | Sepal.Width | Petal.Width | Species
#> -----------------------------------------------------------------
#>         1.00 |            0 |           0 |           0 |  setosa
#>         1.66 |            0 |           0 |           0 |  setosa
#>         2.31 |            0 |           0 |           0 |  setosa
#>         2.97 |            0 |           0 |           0 |  setosa
#>         3.62 |            0 |           0 |           0 |  setosa
#>         4.28 |            0 |           0 |           0 |  setosa
#>         4.93 |            0 |           0 |           0 |  setosa
#>         5.59 |            0 |           0 |           0 |  setosa
#>         6.24 |            0 |           0 |           0 |  setosa
#>         6.90 |            0 |           0 |           0 |  setosa
#> 
#> Maintained constant: Sepal.Length, Sepal.Width, Petal.Width, Species
```

<sup>Created on 2021-05-12 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

##### Multiple targets

By default, `target = "all"`, which means that all combinations are included. One can specify how many variables as he wants using names or column indices. Also, target specifications do work as well. The "column selection" could be further improved if one day we have a smart `data_select()` so that we can use patterns and NSE.
 
``` r
data <- iris[c(3, 4, 5)]

modelbased::visualisation_matrix(data, length = 2) # All targets
#> Visualisation Matrix
#> 
#> Petal.Length | Petal.Width |    Species
#> ---------------------------------------
#>         1.00 |        0.10 |     setosa
#>         6.90 |        0.10 |     setosa
#>         1.00 |        2.50 |     setosa
#>         6.90 |        2.50 |     setosa
#>         1.00 |        0.10 | versicolor
#>         6.90 |        0.10 | versicolor
#>         1.00 |        2.50 | versicolor
#>         6.90 |        2.50 | versicolor
#>         1.00 |        0.10 |  virginica
#>         6.90 |        0.10 |  virginica
#>         1.00 |        2.50 |  virginica
#>         6.90 |        2.50 |  virginica

modelbased::visualisation_matrix(data, target = c("Petal.Length", "Petal.Width"), length = 2)
#> Visualisation Matrix
#> 
#> Petal.Length | Petal.Width | Species
#> ------------------------------------
#>         1.00 |        0.10 |  setosa
#>         6.90 |        0.10 |  setosa
#>         1.00 |        2.50 |  setosa
#>         6.90 |        2.50 |  setosa
#> 
#> Maintained constant: Species

modelbased::visualisation_matrix(data, target = c("Petal.Width = c(0, 1, 2)", "Petal.Length = [5, 7]"), length = 4)
#> Visualisation Matrix
#> 
#> Petal.Width | Petal.Length | Species
#> ------------------------------------
#>           0 |         5.00 |  setosa
#>           1 |         5.00 |  setosa
#>           2 |         5.00 |  setosa
#>           0 |         5.67 |  setosa
#>           1 |         5.67 |  setosa
#>           2 |         5.67 |  setosa
#>           0 |         6.33 |  setosa
#>           1 |         6.33 |  setosa
#>           2 |         6.33 |  setosa
#>           0 |         7.00 |  setosa
#>           1 |         7.00 |  setosa
#>           2 |         7.00 |  setosa
#> 
#> Maintained constant: Species
```

<sup>Created on 2021-05-12 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

Importantly, there's the option to **preserve the range**, so that only "possible" values within each combination of target factors are kept. 

``` r
modelbased::visualisation_matrix(iris, target = c("Petal.Width", "Species"), preserve_range = TRUE)
#> Visualisation Matrix
#> 
#> Petal.Width |    Species | Sepal.Length | Sepal.Width | Petal.Length
#> --------------------------------------------------------------------
#>        0.10 |     setosa |         5.84 |        3.06 |         3.76
#>        0.37 |     setosa |         5.84 |        3.06 |         3.76
#>        1.17 | versicolor |         5.84 |        3.06 |         3.76
#>        1.43 | versicolor |         5.84 |        3.06 |         3.76
#>        1.70 | versicolor |         5.84 |        3.06 |         3.76
#>        1.43 |  virginica |         5.84 |        3.06 |         3.76
#>        1.70 |  virginica |         5.84 |        3.06 |         3.76
#>        1.97 |  virginica |         5.84 |        3.06 |         3.76
#>        2.23 |  virginica |         5.84 |        3.06 |         3.76
#>        2.50 |  virginica |         5.84 |        3.06 |         3.76
#> 
#> Maintained constant: Sepal.Length, Sepal.Width, Petal.Length
```

<sup>Created on 2021-05-13 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

### Models

---

## TO DO

- [x] Fix the methods with the models
- [x] Standardization method (needs https://github.com/easystats/effectsize/pull/324)

### CODE REVIEW

File: NAMESPACE [hunk 1]
```diff
@@ -39,6 +39,7 @@ S3method(visualisation_matrix,polr)
 S3method(visualisation_matrix,stanreg)
 S3method(visualisation_matrix,vector)
 S3method(visualisation_matrix,visualisation_matrix)
+export(.preserve_range)
```
Comment (strengejacke) [inline NAMESPACE:42]: are you exporting an internal here?

Comment (DominiqueMakowski) [reply NAMESPACE:42]: don't judge my WIP code 🙈

**Comment by DominiqueMakowski** [comment]
Any important features from e.g., ggeffects that come to your minds that are currently not covered in the examples above?

**Comment by mattansb** [comment]
<sub>@DominiqueMakowski if you want to port `ggeffects` into `easystats`, there are easier ways....</sub>

**Comment by bwiernik** [comment]
That does raise a question that's been on my mind-- @DominiqueMakowski is there a roadmap for the scope and features of modelbased? The relationship between ggeffects and modelbased is a bit murky, and it would be helpful to see the big picture as to where this package is going. If ggeffects isn't going to be integrated into easystats with a consistent API, etc., it would be helpful to be able to clearly see the hoped-for feature set to guide development.

**Comment by DominiqueMakowski** [comment]
> is there a roadmap for the scope and features of modelbased?

Yeah the scope overlaps a lot with ggeffects. With Daniel we discussed the integration of the latter with easystats a long time ago, and we decided that it'd best to keep it separate because of its different API scheme and its (at least in the name) connectedness with the tidyverse / ggverse.

And since it's a pretty popular package, I suppose many people got used to using it, so it's hard for them to see the (future) advantages of modelbased. <sub>~(and when I say people I obviously mean Mattan 😬)~</sub>

Initially, the main difference was also that modelbased focused on marginal means / contrasts (using emmeans), whereas ggeffects was mostly based on a predict()-like approach. But now both do both so 🤷 Also I would say (Daniel might correct me) that ggeffects is philosophically more focused on plots, which was (initially) not the primary goal of modelbased (I mean, it creates data suited for plotting, but it doesn't make plots on its own).

One could say that modelbased aims to be an easystats-like version of ggeffects, i.e., but with our familiar API, typical features, pipeline and internal usage of the whole easystats machinery.

Because I'm not a user of ggeffects, it's hard for me to see all the overlaps, and because Daniel is obviously not a user of modelbased, it's hard for him too (that's why I need your comments). 

But in any case, having a package devoted to "post-processing / analysis" of models is definitely needed in easystats.

As far as I know, the only feature that really overlaps would be the one mentioned above, i.e. the construction of the visualisation grid. Then the rest is just making predictions on it etc. Because it's the main area of overlap, I want to make sure that it has all the features (and more) that ggeffects or other alternatives offer, so that it has an added value.
 
My goal is first to refactor and simplify the functions a bit (I just wanted to add a feature to visualisation_matrix but here we are I hate myself). As soon as it's done, I'll open an issue to discuss ideas of features/scope evolution ☺️

**Comment by DominiqueMakowski** [comment]
My answer was a bit disorganized haha but the bottom line is that there are a bunch of features that need to be available in easystats, which imo are enough to have a package on their own, now the scope of that package is still to-be-fleshed-out, it could be something *encompassing* these features, instead of defined by them.

Based on its name, the package aims at doing stuff based on models (it's fairly vague I admit 😅), like estimations and such. Now that I think of it though, maybe the extraction and postprocessing of group-specific parameters would be more suited for modelbased than parameters 🤔 especially since Daniel doesn't see the point of that feature haha

**Comment by DominiqueMakowski** [comment]
(also @bwiernik allow me to draw your attention on this feature https://github.com/easystats/see/issues/12 that would be imho a **game changer** for the plotting of differences, unfortunately because of the technical challenge our implementation was a bit wonky - but if we managed to get it stable and robust it would be *awesome*)

**Comment by mattansb** [comment]
> And since it's a pretty popular package, I suppose many people got used to using it, so it's hard for them to see the (future) advantages of modelbased. <sub>~(and when I say people I obviously mean Mattan 😬)~</sub>

Hey, no need so such shade! I only started using `ggeffects` this year! 

<sub>The reason I don't use model based is because I am more comfortable with emmeans...</sub>

I think that if modelbased can cover both the basic uses of emmeans (estimated means, slopes, contrasts) and ggeffects (+ some other things, like plotting not only means, but contrasts as well! and maybe serve as a usable alternative to ggsignif?) it can be a powerful tool for (novice?) users who might find emmeans intimidating...

**Comment by DominiqueMakowski** [comment]
> it can be a powerful tool for (novice?) users who might find emmeans intimidating...

Yeah exactly it's true that the API of emmeans is a bit complex at first

**Comment by bwiernik** [comment]
I certainly find the emmeans API confusing AF and don't like teaching it haha

**Comment by mattansb** [comment]
Noooo emmeans is the best!!! <3 <3 <3 <3 <3

**Comment by DominiqueMakowski** [comment]
Thanks to the recent effectsize update, we can now standardize the predicted matrix (based on the original data), which is superuseful to plot changes in terms of modifications of SD.

``` r
pred <- modelbased::estimate_expectation(lm(mpg ~ wt, data = mtcars))
effectsize::standardize(pred, include_response = FALSE)
#> Variable predicted: mpg
#> 
#> wt    | Predicted |   SE |         95% CI
#> -----------------------------------------
#> -1.74 |     29.20 | 1.09 | [26.96, 31.43]
#> -1.30 |     26.88 | 0.89 | [25.06, 28.70]
#> -0.85 |     24.55 | 0.71 | [23.10, 26.01]
#> -0.41 |     22.23 | 0.58 | [21.04, 23.42]
#> 0.03  |     19.91 | 0.54 | [18.81, 21.01]
#> 0.48  |     17.59 | 0.60 | [16.36, 18.81]
#> 0.92  |     15.26 | 0.74 | [13.76, 16.77]
#> 1.37  |     12.94 | 0.92 | [11.06, 14.82]
#> 1.81  |     10.62 | 1.13 | [ 8.32, 12.92]
#> 2.26  |      8.30 | 1.35 | [ 5.55, 11.05]
```

<sup>Created on 2021-05-21 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>


It aims at replacing the [older `standardize` argument](https://easystats.github.io/modelbased/articles/visualisation_matrix.html#visualising-an-interaction-between-two-numeric-variables-three-way-interaction) directly in the visualisation_matrix call, which made the whole implementation more complicated and clunky.


When `include_response` is TRUE, the function will also loop through the predicted column, as well as the CI_low and CI_high, to standardize it according to the original response variable:

https://github.com/easystats/modelbased/blob/add8c16ee6d169417d03051f7a182eef0ec41c51/R/standardize_methods.R#L25-L32

However, how to also standardize the `SE` column 🤔 is it possible? would it be just to divide it by the SD of the response of something like that? @bwiernik @mattansb @strengejacke

**Comment by bwiernik** [comment]
Depends on the type of model and the sources of uncertainty you care about. 

Dividing by the response SD like this works for standardizing linear model estimates, SEs, and confidence intervals, but not necessarily GLMs.

Scaling the predictor variables like this should be fine I think. 

Additionally, this approach does not account for the uncertainty in the SD estimate used to standardize the variables, so the SE and CI will be somewhat too narrow. Most treatments of standardized coefficients and predictions ignore this point however. 

Presentation-wise, I think the meaning of the resulting table is a little unclear. It's not obvious at first sight I don't think that the values in `wt` are the mean and +/- several SDs. I think perhaps changing the reported values to 0, +1 SD, +2 SD, etc. and changing the name of the columns to `z_wt` or keeping the name but printing as `wt (standardized)` would be clearer.

**Comment by DominiqueMakowski** [comment]
> Depends on the type of model

Yes this is only applied if the model is linear

> Presentation-wise, I think the meaning of the resulting table is a little unclear

In the past version (where standardized parameters were computed directly by enabling an option), all columns were indeed renamed with the `Std_` colname. But now, because it is done a posterior via the `standardize()` method, it seemed kinda weird to me to modify the name. After all, when running `standardize(mtcars)`, you wouldn't expect names to be changed...

So for now I added `(standardized)` to the title of the table:

``` r
effectsize::standardize(modelbased::estimate_expectation(lm(mpg ~ vs, data=mtcars)))
#> Model-based Expectation (standardized)
#> 
#> vs    | Predicted |   SE |         95% CI
#> -----------------------------------------
#> -0.87 |     -0.58 | 1.08 | [-0.94, -0.21]
#> -0.65 |     -0.43 | 0.97 | [-0.76, -0.10]
#> -0.43 |     -0.28 | 0.88 | [-0.58,  0.02]
#> -0.21 |     -0.14 | 0.83 | [-0.42,  0.14]
#> 0.01  |  9.15e-03 | 0.81 | [-0.27,  0.28]
#> 0.23  |      0.16 | 0.83 | [-0.13,  0.44]
#> 0.45  |      0.30 | 0.89 | [ 0.00,  0.60]
#> 0.68  |      0.45 | 0.98 | [ 0.12,  0.78]
#> 0.90  |      0.59 | 1.09 | [ 0.22,  0.97]
#> 1.12  |      0.74 | 1.22 | [ 0.33,  1.16]
#> 
#> Variable predicted: mpg
#> Predictors modulated: vs
```

<sup>Created on 2021-05-22 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

But I'm not sure either whether it's the best solution. What do you guys think? (also pinging @IndrajeetPatil)

**Comment by DominiqueMakowski** [comment]
What started with "just adding a small feature" ended up by ***not*** adding said feature, but instead rewriting the whole package. 
Anyway, it's done now, and modelbased should be more flexible, powerful and maintainable than ever. Big changes include:

- General overhaul of the package. 
- Entire refactoring of `visualisation_matrix()` with added features 
- Option of standardizing/unstandardizing predictions, contrasts and means is now available via `standardize()` instead of via options.
- Introduction of `model_emmeans()` as a wrapper to easily create `emmeans` objects.
- `estimate_smooth()` transformed into `describe_nonlinear()` and made more explicit.   

Now we can start thinking and discussing the features paths of development ☺️

**Pull request merged**